### ISSUE TITLE
USB Camera, Add device with by-id

### ISSUE THREAD

**Issue by charredchar** [AlexxIT/go2rtc#1237]
I am running go2rtc on an Ubuntu 22.04 device with multiple USB cameras plugged in.
When using the `/dev/video#` option provided by FFMPEG in the Add tab I have an issue where the cameras will randomly change when the system reboots.
I attempted to set it to the `/dev/v4l/by-id/` property instead but I always receive this error.

```
undefined error=streams: exec/rtsp
[video4linux2,v4l2 @ 0x7039b849b600] Cannot open video device /dev/v4l/by-id/usb-046d_HD_Webcam_C525_B994AB20-video-index0: No such file or directory
[in#0 @ 0x7039b85918c0] Error opening input: No such file or directory
Error opening input file /dev/v4l/by-id/usb-046d_HD_Webcam_C525_B994AB20-video-index0.
Error opening input files: No such file or directory
caller=github.com/AlexxIT/go2rtc/internal/hls/hls.go:76
```

This is an example of my config with one camera as `by-id` and the other by `video#`. Only the IMX219 works in this example.

```
streams:
  C525: ffmpeg:device?video=/dev/v4l/by-id/usb-046d_HD_Webcam_C525_B994AB20-video-index0&input_format=yuyv422&video_size=1184x656#video=h264#hardware
  IMX219: ffmpeg:device?video=/dev/video0&input_format=yuyv422&video_size=1024x768#video=h264#hardware
```

**Comment by mitchhh22** [comment]
I have 4 usb cameras plugged into a usb hub. I run into this same error because the /dev/video# constantly changes and no way to assign it to be fixed. Using the path or id yields the same "No such file or directory" and creating a new symlink doesnt work either.

**Comment by mitchhh22** [comment]
@charredchar did you ever figure out a solution or workaround here?

**Comment by Devirex** [comment]
I had the same issue, found this workarround:

```yaml 
streams:
  cam1: echo:bash mycam.sh
````

mycam.sh
```bash
cam=$(realpath /dev/v4l/by-id/usb-icSpring_WebCamera_20240603201233-video-index0)
echo 'exec:ffmpeg -re -i '$cam' -f rtsp {output}'
```
this is just a quick example, i think you can echo any command you usually would be able to write directly into the config.

**Comment by lindsaymarkward** [comment]
This has been an issue for me with 2 USB cameras on a Raspberry Pi 4. I'd love a solution builtin. Thanks!

**Comment by flxinxout** [comment]
UP :) 
I have the same issue. I executed the previous answer but I get this:

```
sudo ffmpeg -re -i /dev/video0 -f rtsp {output}
ffmpeg version 6.1.1-3ubuntu5 Copyright (c) 2000-2023 the FFmpeg developers
  built with gcc 13 (Ubuntu 13.2.0-23ubuntu3)
  configuration: --prefix=/usr --extra-version=3ubuntu5 --toolchain=hardened --libdir=/usr/lib/aarch64-linux-gnu --incdir=/usr/include/aarch64-linux-gnu --arch=arm64 --enable-gpl --disable-stripping --disable-omx --enable-gnutls --enable-libaom --enable-libass --enable-libbs2b --enable-libcaca --enable-libcdio --enable-libcodec2 --enable-libdav1d --enable-libflite --enable-libfontconfig --enable-libfreetype --enable-libfribidi --enable-libglslang --enable-libgme --enable-libgsm --enable-libharfbuzz --enable-libmp3lame --enable-libmysofa --enable-libopenjpeg --enable-libopenmpt --enable-libopus --enable-librubberband --enable-libshine --enable-libsnappy --enable-libsoxr --enable-libspeex --enable-libtheora --enable-libtwolame --enable-libvidstab --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libx265 --enable-libxml2 --enable-libxvid --enable-libzimg --enable-openal --enable-opencl --enable-opengl --disable-sndio --enable-libdc1394 --enable-libdrm --enable-libiec61883 --enable-chromaprint --enable-frei0r --enable-ladspa --enable-libbluray --enable-libjack --enable-libpulse --enable-librabbitmq --enable-librist --enable-libsrt --enable-libssh --enable-libsvtav1 --enable-libx264 --enable-libzmq --enable-libzvbi --enable-lv2 --enable-sdl2 --enable-libplacebo --enable-librav1e --enable-pocketsphinx --enable-librsvg --enable-libjxl --enable-shared
  libavutil      58. 29.100 / 58. 29.100
  libavcodec     60. 31.102 / 60. 31.102
  libavformat    60. 16.100 / 60. 16.100
  libavdevice    60.  3.100 / 60.  3.100
  libavfilter     9. 12.100 /  9. 12.100
  libswscale      7.  5.100 /  7.  5.100
  libswresample   4. 12.100 /  4. 12.100
  libpostproc    57.  3.100 / 57.  3.100
Input #0, video4linux2,v4l2, from '/dev/video0':
  Duration: N/A, start: 1702.007045, bitrate: 147456 kb/s
  Stream #0:0: Video: rawvideo (YUY2 / 0x32595559), yuyv422, 640x480, 147456 kb/s, 30 fps, 30 tbr, 1000k tbn
Stream mapping:
  Stream #0:0 -> #0:0 (rawvideo (native) -> mpeg4 (native))
Press [q] to stop, [?] for help
[tcp @ 0xaaaadd3c2fe0] Connection to tcp://:554?timeout=0 failed: Connection refused
[out#0/rtsp @ 0xaaaadd2ea730] Could not write header (incorrect codec parameters ?): Connection refused
Error while filtering: Connection refused
[out#0/rtsp @ 0xaaaadd2ea730] Nothing was written into output file, because at least one of its streams received no packets.
frame=    0 fps=0.0 q=0.0 Lsize=       0kB time=N/A bitrate=N/A speed=N/A    
Conversion failed!

```

**Comment by jheidel** [comment]
This seems to work for me. Perhaps something has changed since this issue was filed? Maybe it's only applicable when adding devices through the API?

---

I've confirmed this config works where I'm passing a symlinked v4l device path to the `video` parameter:

```yaml
streams:
  camera:   ffmpeg:device?video=/dev/v4l/by-id/usb-Microsoft_Front_LifeCam-video-index0&video_size=1280x720#video=h264
```

When I load the stream, I can see that `ffmpeg` is started with the v4l symlink input device as expected (and playback works fine):

```bash
$ ps axu | grep ffmpeg
jeff        1474 86.1  3.5 967452 139636 ?       Sl   09:51   0:06 ffmpeg -hide_banner -v error -f v4l2 -video_size 1280x720 -i /dev/v4l/by-id/usb-Microsoft_Front_LifeCam-video-index0 -c:v libx264 -g 50 -profile:v high -level:v 4.1 -preset:v superfast -tune:v zerolatency -pix_fmt:v yuv420p -an -user_agent ffmpeg/go2rtc -rtsp_transport tcp -f rtsp rtsp://127.0.0.1:8554/4b98c3b2a62faeb3c96114d6338cb129
```

This behavior lines up with the implementation. The `ffmpeg -i` flag is configured using `indexToItem`:
https://github.com/AlexxIT/go2rtc/blob/34b103bbcba512a07fb1ea8a3c219ae31d2201bf/internal/ffmpeg/device/device_unix.go#L30
and that helper will simply pass through the `index` if it is not found in the initialized video devices
https://github.com/AlexxIT/go2rtc/blob/34b103bbcba512a07fb1ea8a3c219ae31d2201bf/internal/ffmpeg/device/devices.go#L41-L46

In other words, if you don't provide a valid index, it simply passes the value through to ffmpeg.
